keylime/cert_utils.py

```diff
@@ -1,13 +1,133 @@
-    return load_der_public_key(encoder.encode(der509["tbsCertificate"]["subjectPublicKeyInfo"]))
+
+# Here we provide some helpers that use pyasn1 to parse the certificates
+# when parsing them with python-cryptography fails, and in this case, we
+# try to read the parsed certificate again into python-cryptograhy.
+
+logger = keylime_logging.init_logging("cert_utils")
+
+
+class UserError(Exception):
+    pass
+
+
+def x509_der_cert(der_cert_data):
+    if isinstance(der_cert_data, str):
+        der_cert_data = der_cert_data.encode("utf-8")
+    try:
+        cert = cert = x509.load_der_x509_certificate(data=der_cert_data, backend=default_backend())
+        return cert
+    except Exception:
+        pyasn1_cert = decoder.decode(der_cert_data, asn1Spec=rfc2459.Certificate())[0]
+        return x509.load_der_x509_certificate(data=encoder.encode(pyasn1_cert), backend=default_backend())
+
+
+def x509_pem_cert(pem_cert_data):
+    if isinstance(pem_cert_data, (bytes, bytearray)):
+        pem_cert_data = pem_cert_data.decode("utf-8")
+    try:
+        cert = x509.load_pem_x509_certificate(data=pem_cert_data, backend=default_backend())
+        return cert
+    except Exception:
+        return x509_der_cert(pem.readPemFromFile(io.StringIO(pem_cert_data)))
+
+
+def verify_ek(ekcert, tpm_cert_store=config.get("tenant", "tpm_cert_store")):
+    """Verify that the provided EK certificate is signed by a trusted root
+    :param ekcert: The Endorsement Key certificate in DER format
+    :returns: True if the certificate can be verified, false otherwise
+    """
+    try:
+        ek509 = x509_der_cert(ekcert)
+
+        trusted_certs = tpm_ek_ca.cert_loader(tpm_cert_store)
+        for cert_file, pem_cert in trusted_certs.items():
+            signcert = x509_pem_cert(pem_cert)
+            if ek509.issuer != signcert.subject:
+                continue
+
+            signcert_pubkey = signcert.public_key()
+            try:
+                if isinstance(signcert_pubkey, RSAPublicKey):
+                    signcert_pubkey.verify(
+                        ek509.signature,
+                        ek509.tbs_certificate_bytes,
+                        padding.PKCS1v15(),
+                        ek509.signature_hash_algorithm,
+                    )
+                elif isinstance(signcert_pubkey, EllipticCurvePublicKey):
+                    signcert_pubkey.verify(
+                        ek509.signature,
+                        ek509.tbs_certificate_bytes,
+                        ec.ECDSA(ek509.signature_hash_algorithm),
+                    )
+                else:
+                    logger.warning("Unsupported public key type: %s", type(signcert_pubkey))
+                    continue
+            except crypto_exceptions.InvalidSignature:
+                continue
+
+            logger.debug("EK cert matched cert: %s", cert_file)
+            return True
+    except Exception as e:
```

Comment:
This can hide problems in the certification load under a misleading error message.

The `try` should be moved early, surrounding only `tpm_ek_ca.cert_loader`, because as it is now it is also catching any exception in the PEM / DER loads, that does not indicate any issue with the TPM.

Reply:
I changed to be in line with your comment. Let me know it if is better now, please.